File: week-02/day-2/src/StudentCounter.java
import java.util.*;

public class StudentCounter {

  public static void main(String... args) {
    List<Map<String, Object>> listOfMaps = new ArrayList<>();

    // Given this list of hash maps...

    Map<String, Object> row0 = new HashMap<>();
    row0.put("name", "Theodor");
    row0.put("age", 9.5);
    row0.put("candies", 2);
    listOfMaps.add(row0);

    Map<String, Object> row1 = new HashMap<>();
    row1.put("name", "Paul");
    row1.put("age", 10);
    row1.put("candies", 1);
    listOfMaps.add(row1);

    Map<String, Object> row2 = new HashMap<>();
    row2.put("name", "Mark");
    row2.put("age", 7);
    row2.put("candies", 3);
    listOfMaps.add(row2);

    Map<String, Object> row3 = new HashMap<>();
    row3.put("name", "Peter");
    row3.put("age", 12);
    row3.put("candies", 5);
    listOfMaps.add(row3);

    Map<String, Object> row4 = new HashMap<>();
    row4.put("name", "Olaf");
    row4.put("age", 12);
    row4.put("candies", 7);
    listOfMaps.add(row4);

    Map<String, Object> row5 = new HashMap<>();
    row5.put("name", "George");
    row5.put("age", 3);
    row5.put("candies", 2);
    listOfMaps.add(row5);

    ArrayList<Map> rows = new ArrayList<>();
    rows.add(row0);
    rows.add(row1);
    rows.add(row2);
    rows.add(row3);
    rows.add(row4);
    rows.add(row5);

    // Display the following things:
    //  - The names of students who have more than 4 candies
    System.out.println("Students who have more than 4 candies: ");
    for (Map<String, Integer> row : rows) {
      if (row.get("candies") > 4) {
        System.out.println(row.get("name"));
      }
    }

    //  - The sum of the age of people who have less than 5 candies
    double sumAge = 0;
    System.out.println("The sum of the age of people who have less than 5 candies: ");
    for (Map<String, Object> row : rows) {
      if ((int) (row.get("candies")) < 5) {
        try {
          sumAge += ((Integer) row.get("age")).doubleValue();
        } catch (ClassCastException e) {
          sumAge += ((Double) row.get("age"));
        }
      }
    }
    System.out.println(sumAge);
  }
}

File: python_solutions/week-02/day-2/data_structures/student_counter.py
students = [
    {'name': 'Theodor', 'age': 3, 'candies': 2},
    {'name': 'Paul', 'age': 9.5, 'candies': 2},
    {'name': 'Mark', 'age': 12, 'candies': 5},
    {'name': 'Peter', 'age': 7, 'candies': 3},
    {'name': 'Olaf', 'age': 12, 'candies': 7},
    {'name': 'George', 'age': 10, 'candies': 1}
]


# create a function that takes a list of students and prints:
# - how many candies are owned by students altogether

def candy_counter(students: list):
    return sum(student.get("candies") for student in students)


# create a function that takes a list of students and prints:
# - The sum of the age of people who have less than 5 candies

def age_sum(students: list):
    students = filter(lambda student: student.get("candies") < 5, students)
    return sum(student.get("age") for student in students)


if __name__ == '__main__':
    print(candy_counter(students))
    print(age_sum(students))
